decode.go

```diff
@@ -178,19 +182,44 @@ func findFirstMultiByteChar(data string) int {
 	return len(data)
 }
 
-func (d *Decoder) readLine(v reflect.Value) (err error, ok bool) {
-	line, err := d.data.ReadString('\n')
-	if err != nil && err != io.EOF {
-		return err, false
+func (d *Decoder) SetSeparator(separator string) {
+	d.separator = separator
+}
+
+func (d Decoder) Separator() string {
+	if d.separator != "" {
+		return d.separator
 	}
-	if err == io.EOF {
-		d.done = true
 
-		if len(line) <= 0 || line[0] == '\n' {
-			// skip last empty lines
-			return nil, false
-		}
+	return "\n"
+}
+
+func (d *Decoder) Scan(data []byte, atEOF bool) (advance int, token []byte, err error) {
```

Comment:
Why implement the custom scan function? I think `data.ReadString(d.lineTerminator)` would work fine.

Reply:
@ianlopshire `ReadString` only takes a single byte as argument: https://golang.org/pkg/bufio/#Reader.ReadString

So in the case of "\r\n" this wouldn't work